source/uwp/SharedRenderer/lib/AdaptiveCarouselRenderer.cpp

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "pch.h"
+#include "ActionHelpers.h"
+#include "AdaptiveCarouselRenderer.h"
+#include "AdaptiveCarouselRenderer.g.cpp"
+#include "AdaptiveRenderArgs.h"
+#include "WholeItemsPanel.h"
+#include "XamlBuilder.h"
+#include <windows.ui.xaml.data.h>
+
+namespace winrt::AdaptiveCards::Rendering::Xaml_Rendering::implementation
+{
+    winrt::UIElement AdaptiveCarouselRenderer::Render(
+        winrt::AdaptiveCards::ObjectModel::Xaml_Rendering::IAdaptiveCardElement const& element,
+        winrt::AdaptiveCards::Rendering::Xaml_Rendering::AdaptiveRenderContext const& context,
+        winrt::AdaptiveCards::Rendering::Xaml_Rendering::AdaptiveRenderArgs const& renderArgs)
+    {
+        auto carousel = element.as<winrt::IAdaptiveCarousel>();
+        auto styledCollection = element.as<winrt::IAdaptiveContainerBase>();
+
+        StackPanel stackPanel{};
+        FlipView carouselUI{};
+        bool loopEnabled{false};
+
+        if (carousel.AutoLoop())
+        {
+            loopEnabled = carousel.AutoLoop().GetBoolean();
+        }
+
+// WinUI3 is currently the primary target for Carousel.
+// Carousel is compatible with UWP; however, pagination is not supported.
```

Comment:
I'm curious why, does `PipsPager` not work in UWP?:)

Reply:
It requires Window UI Library 2. We can add it when's needed. https://microsoft.visualstudio.com/AdaptiveCards/_build/results?buildId=89472031&view=artifacts&pathAsName=false&type=publishedArtifacts